### PR TITLE
[debug] adding OT_ASSERT() macro (allow disabling asserts)

### PR DESCRIPTION
**Pull request by abtink**
This PR contains two related commits:

**[debug] adding OT_ASSERT() macro**

This commit adds a new OpenThread specific `OT_ASSERT()` macro
along with a new config OPENTHREAD_CONFIG_ASSERT_ENABLE option which
allows the assert to be enabled or disabled. This allows a release
build to consider disabling the assert reducing code size.

**[travis] add new simul platform build with asserts disabled**

### CODE REVIEW

File: src/core/common/debug.hpp [hunk 1]
```diff
@@ -71,23 +71,25 @@
...
+            while (1)   \
+            {           \
+            }           \
+        }               \
+    } while (0)
 
-#ifdef assert
-#undef assert
-#endif
+#endif // #elif OPENTHREAD_CONFIG_PLATFORM_ASSERT_MANAGEMENT
```
Comment (jwhui) [inline src/core/common/debug.hpp]: nit: we generally do not include the `#elif` in the comment

**Comment by bukepo** [comment]
Could we mimic the standard `assert()` behavior? i.e. use the same name and disabled by the same definition `NDEBUG`?

File: src/core/common/debug.hpp [hunk 2]
```diff
@@ -73,21 +73,23 @@
...
+            }           \
+        }               \
     } while (0)
 
-#endif
+#endif // OPENTHREAD_CONFIG_PLATFORM_ASSERT_MANAGEMENT
+
+#else
+
+#define OT_ASSERT(cond)
```
Comment (simonlingoogle) [inline src/core/common/debug.hpp]:
```suggestion
#define OT_ASSERT(cond) ((void)0)
```
otherwise some lint program may complain about empty statements.

**Comment by abtink** [comment]
> Could we mimic the standard assert() behavior? i.e. use the same name and disabled by the same definition NDEBUG?

The main reason for adding an OT specific assert `OT_ASSERT()` macro (instead of the `assert()`) is to avoid any macro/function name collisions with other definitions from third-party or platform layer code (e.g., the issue that was fixed from https://github.com/openthread/openthread/pull/4329 where platform layer asserts (intended for mbed-tls module) were being used instead). 

With an `OT_ASSERT()` macro, it is clear it is OT specific assert and we allow platform layer to customize the behavior (using `OPENTHREAD_CONFIG_PLATFORM_ASSERT_MANAGEMENT`) on how to  handle asserts in OT core may be even differently from asserts from other modules.

Thoughts?

File: src/core/common/debug.hpp (hunk 2 above)
Comment (abtink) [reply src/core/common/debug.hpp]: We actually do use a lot of empty macros for OT logging (`otLogDebg`, etc.) already.
I would suggest we keep this as is and use empty macro here as well. We can change them all if/when there is an issue with using empty macros.

**Pull request merged**